beacon_node/http_api/tests/broadcast_validation_tests.rs

```diff
@@ -1395,3 +1400,366 @@ pub async fn blinded_equivocation_full_pass() {
+        .harness
+        .chain
+        .block_is_known_to_fork_choice(&block.canonical_root()));
+
+    // Post the block *and* all blobs to the HTTP API.
+    let response: Result<(), eth2::Error> = tester
+        .client
+        .post_beacon_blocks_v2(
+            &PublishBlockRequest::new(block.clone(), Some((kzg_proofs, blobs))),
+            validation_level,
+        )
+        .await;
+
+    // This should result in the block being fully imported.
+    response.unwrap();
+    assert!(tester
+        .harness
+        .chain
+        .block_is_known_to_fork_choice(&block.canonical_root()));
+}
+
+/// This test checks that an HTTP POST request with the block succeeds with a 200 response
+/// if just the blobs have already been seen on gossip.
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+pub async fn blobs_seen_on_gossip_without_block_and_no_http_blobs() {
+    let validation_level: Option<BroadcastValidation> = Some(BroadcastValidation::Gossip);
+
+    // Validator count needs to be at least 32 or proposer boost gets set to 0 when computing
+    // `validator_count // 32`.
+    let validator_count = 64;
+    let num_initial: u64 = 31;
+    let spec = ForkName::latest().make_genesis_spec(E::default_spec());
+    let tester = InteractiveTester::<E>::new(Some(spec), validator_count).await;
+
+    // Create some chain depth.
+    tester.harness.advance_slot();
+    tester
+        .harness
+        .extend_chain(
+            num_initial as usize,
+            BlockStrategy::OnCanonicalHead,
+            AttestationStrategy::AllValidators,
+        )
+        .await;
+    tester.harness.advance_slot();
+
+    let slot_a = Slot::new(num_initial);
+    let slot_b = slot_a + 1;
+
+    let state_a = tester.harness.get_current_state();
+    let ((block, blobs), _) = tester.harness.make_block(state_a, slot_b).await;
+    let (kzg_proofs, blobs) = blobs.expect("should have some blobs");
+    assert!(!blobs.is_empty());
+
+    // Simulate the blobs being seen on gossip.
+    for (i, (kzg_proof, blob)) in kzg_proofs
+        .clone()
+        .into_iter()
+        .zip(blobs.clone())
+        .enumerate()
+    {
+        let sidecar = Arc::new(BlobSidecar::new(i, blob, &block, kzg_proof).unwrap());
+        let gossip_blob =
+            GossipVerifiedBlob::new(sidecar, i as u64, &tester.harness.chain).unwrap();
+        tester
+            .harness
+            .chain
+            .process_gossip_blob(gossip_blob, || panic!("should not publish block yet"))
+            .await
+            .unwrap();
+    }
+
+    // It should not yet be added to fork choice because the block has not been seen.
+    assert!(!tester
+        .harness
+        .chain
+        .block_is_known_to_fork_choice(&block.canonical_root()));
+
+    // Post just the block to the HTTP API.
+    let response: Result<(), eth2::Error> = tester
+        .client
+        .post_beacon_blocks_v2(
+            &PublishBlockRequest::new(block.clone(), None),
```

Comment:
From Deneb the request is assumed to be `PublishBlockRequest::BlockContents`:
- ssz for Deneb request is always decoded to `BlockContents`: https://github.com/sigp/lighthouse/blob/79680c886dcdd1577edbd3a93f920df221cafc8c/common/eth2/src/types.rs#L1825 
- `TryFrom` impl does not allow constructing from only signed block: https://github.com/sigp/lighthouse/blob/79680c886dcdd1577edbd3a93f920df221cafc8c/common/eth2/src/types.rs#L1909-L1911

Although I think this code path is still possible from VC sending json body.

Reply:
I'm wonder under what scenario does this happen? or is this test to make sure we cover all cases?

Reply:
This is meant to test the scenario where the caller POSTs a block without blobs, but you're right that it would be better to do that with a `PublishBlockRequest::BlockContents` where the list of blobs is empty. I will update accordingly.

Reply:
Test updated